In [1]:
#importing libraries

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt # we only need pyplot
%matplotlib inline
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from datetime import datetime

In [2]:
music = pd.read_csv("cleaned_music.csv")
music

Unnamed: 0  index  popularity  acousticness  danceability  duration_ms  \
0               0      1        31.0       0.01270         0.622     218293.0   
1               1      2        28.0       0.00306         0.620     215613.0   
2               2      3        34.0       0.02540         0.774     166875.0   
3               3      4        32.0       0.00465         0.638     222369.0   
4               4      6        46.0       0.02890         0.572     214408.0   
...           ...    ...         ...           ...           ...          ...   
40555       40555  49999        56.0       0.13300         0.849     237667.0   
40556       40556  50001        72.0       0.15700         0.709     251860.0   
40557       40557  50002        51.0       0.00597         0.693     189483.0   
40558       40558  50003        65.0       0.08310         0.782     262773.0   
40559       40559  50004        67.0       0.10200         0.862     267267.0   

       energy  instrumentalness key  liveness  loudness   mode  speechiness  \
0       0.890          0.950000   D     0.124    -7.043  Minor       0.0300   
1       0.755          0.011800  G#     0.534    -4.617  Major       0.0345   
2       0.700          0.002530  C#     0.157    -4.498  Major       0.2390   
3       0.587          0.909000  F#     0.157    -6.266  Major       0.0413   
4       0.803          0.000008   B     0.106    -4.294  Major       0.3510   
...       ...               ...  ..       ...       ...    ...          ...   
40555   0.660          0.000008   C     0.296    -7.195  Major       0.0516   
40556   0.362          0.000000   B     0.109    -9.814  Major       0.0550   
40557   0.763          0.000000   D     0.143    -5.443  Major       0.1460   
40558   0.472          0.000000   G     0.106    -5.016  Minor       0.0441   
40559   0.642          0.000000  F#     0.272   -13.652  Minor       0.1010   

         tempo  valence music_genre  
0      115.002    0.531  Electronic  
1      127.994    0.333  Electronic  
2      128.014    0.270  Electronic  
3      145.036    0.323  Electronic  
4      149.995    0.230  Electronic  
...        ...      ...         ...  
40555   99.988    0.629     Hip-Hop  
40556  122.043    0.113     Hip-Hop  
40557  131.079    0.395     Hip-Hop  
40558   75.886    0.354     Hip-Hop  
40559   99.201    0.765     Hip-Hop  

[40560 rows x 16 columns]

In [3]:
music = music.drop(["Unnamed: 0"], axis = 1)


From the table, unknown column "Unnamed: 0" spotted. Hence, removed.

# 4) Machine Learnning

We will use 2 predictive models to discover possible patterns in our dataset and make predictions based on them.

#### 4a) Encode categorical data

Most Machine Learning algorithms cannot work with strings; they expect to see numerical values in each feature. It is preferable to one-hot encode categorical features (those having string values), i.e., to expand number of features so that values for each class are placed in a separate column. Where the sample correspond to the given class, it gets 1 in the respective feature; all others get zeroes. Instead, categorical values in this exercise are encoded with number representation. It assigns a number (integer) to each class. Thus, 5 distinctive classes in a categorical variable will get values between 0 and 4. Therefore, an encoder is instantiated for each categorical variable. The labels (i.e., the music genre) are not encoded, which will make interpretation easier and more understandable. Tree-based algorithms accept string labels.

Encoding key feature

In [5]:
music['key'].replace(['A','A#','B','C','C#','D','D#','E','F','F#','G','G#'],
                        [0,1,2,3,4,5,6,7,8,9,10,11], inplace=True)
music.head()

index  popularity  acousticness  danceability  duration_ms  energy  \
0      1        31.0       0.01270         0.622     218293.0   0.890   
1      2        28.0       0.00306         0.620     215613.0   0.755   
2      3        34.0       0.02540         0.774     166875.0   0.700   
3      4        32.0       0.00465         0.638     222369.0   0.587   
4      6        46.0       0.02890         0.572     214408.0   0.803   

   instrumentalness  key  liveness  loudness   mode  speechiness    tempo  \
0          0.950000    5     0.124    -7.043  Minor       0.0300  115.002   
1          0.011800   11     0.534    -4.617  Major       0.0345  127.994   
2          0.002530    4     0.157    -4.498  Major       0.2390  128.014   
3          0.909000    9     0.157    -6.266  Major       0.0413  145.036   
4          0.000008    2     0.106    -4.294  Major       0.3510  149.995   

   valence music_genre  
0    0.531  Electronic  
1    0.333  Electronic  
2    0.270  Electronic  
3    0.323  Electronic  
4    0.230  Electronic

Encoding mode feature

In [6]:
music['mode'].replace(['Major','Minor'],
                        [0,1], inplace=True)
music.head()

index  popularity  acousticness  danceability  duration_ms  energy  \
0      1        31.0       0.01270         0.622     218293.0   0.890   
1      2        28.0       0.00306         0.620     215613.0   0.755   
2      3        34.0       0.02540         0.774     166875.0   0.700   
3      4        32.0       0.00465         0.638     222369.0   0.587   
4      6        46.0       0.02890         0.572     214408.0   0.803   

   instrumentalness  key  liveness  loudness  mode  speechiness    tempo  \
0          0.950000    5     0.124    -7.043     1       0.0300  115.002   
1          0.011800   11     0.534    -4.617     0       0.0345  127.994   
2          0.002530    4     0.157    -4.498     0       0.2390  128.014   
3          0.909000    9     0.157    -6.266     0       0.0413  145.036   
4          0.000008    2     0.106    -4.294     0       0.3510  149.995   

   valence music_genre  
0    0.531  Electronic  
1    0.333  Electronic  
2    0.270  Electronic  
3    0.323  Electronic  
4    0.230  Electronic

#### 4b) Exploration and Visualization

Perform basic statistical exploration and visualization on the Train Set.

We will start by setting up a Multi-Variate Classification problem.

In [ ]:
# Extract Response and Predictors
y = pd.DataFrame(music["music_genre"])
predictors = ["popularity","acousticness","danceability","duration_ms","energy","instrumentalness","key","liveness","loudness","mode","speechiness","tempo","valence"]
X = pd.DataFrame(music[predictors]) 

In [ ]:
# Import train_test_split from sklearn
from sklearn.model_selection import train_test_split

# Split the Dataset into Train and Test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25)

# Check the sample sizes
print("Train Set :", y_train.shape, X_train.shape)
print("Test Set  :", y_test.shape, X_test.shape)

In [ ]:
# Summary Statistics for Response
y_train["music_genre"].value_counts()

In [ ]:
# Summary Statistics for Predictors
X_train.describe()

In [ ]:
# Draw the distribution of Response
sns.catplot(y = "music_genre", data = y_train, kind = "count")

## 4c) Classification Tree

In [ ]:
# Import DecisionTreeClassifier model from Scikit-Learn
from sklearn.tree import DecisionTreeClassifier

# Decision Tree using Train Data
dectree = DecisionTreeClassifier(max_depth = 3)  # create the decision tree object
dectree.fit(X_train, y_train)                    # train the decision tree model
y_predict = dectree.fit(X_train,y_train).predict(X_test)

In [ ]:
# Plot the trained Decision Tree
from sklearn.tree import plot_tree

f = plt.figure(figsize=(30,20))
plot_tree(dectree, filled=True, rounded=True, fontsize = 7, 
          feature_names= X_train.columns, 
          class_names= ["Classical","Jazz","Country","Rock","Hip-Hop","Rap","Anime","Alternative","Blues","Electronic"])

## 4d) Confusion Matrix

#### We start by Finding Optimal Depth via K-fold Cross-Validation
We use a range of tree depths to evaluate and to plot the estimated performance +/- 2 standard deviations for each depth using K-fold cross validation. We provide a Python code that can be used in any situation, where you want to tune your decision tree given a predictor tensor X and labels Y. The code includes the training set performance in the plot, while scaling the y-axis to focus on the cross-validation performance.

In [ ]:
# Import essential models and functions
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.model_selection import cross_val_score
import numpy as np
import matplotlib.pyplot as plt

# function for fitting trees of various depths on the training data using cross-validation
def run_cross_validation_on_trees(X, y, tree_depths, cv=5, scoring='accuracy'):
    cv_scores_list = []
    cv_scores_std = []
    cv_scores_mean = []
    accuracy_scores = []
    for depth in tree_depths:
        tree_model = DecisionTreeClassifier(max_depth=depth)
        cv_scores = cross_val_score(tree_model, X, y, cv=cv, scoring=scoring)
        cv_scores_list.append(cv_scores)
        cv_scores_mean.append(cv_scores.mean())
        cv_scores_std.append(cv_scores.std())
        accuracy_scores.append(tree_model.fit(X, y).score(X, y))
    cv_scores_mean = np.array(cv_scores_mean)
    cv_scores_std = np.array(cv_scores_std)
    accuracy_scores = np.array(accuracy_scores)
    return cv_scores_mean, cv_scores_std, accuracy_scores
  
# function for plotting cross-validation results
def plot_cross_validation_on_trees(depths, cv_scores_mean, cv_scores_std, accuracy_scores, title):
    fig, ax = plt.subplots(1,1, figsize=(15,5))
    ax.plot(depths, cv_scores_mean, '-o', label='mean cross-validation accuracy', alpha=0.9)
    ax.fill_between(depths, cv_scores_mean-2*cv_scores_std, cv_scores_mean+2*cv_scores_std, alpha=0.2)
    ylim = plt.ylim()
    ax.plot(depths, accuracy_scores, '-*', label='train accuracy', alpha=0.9)
    ax.set_title(title, fontsize=16)
    ax.set_xlabel('Tree depth', fontsize=14)
    ax.set_ylabel('Accuracy', fontsize=14)
    ax.set_ylim(ylim)
    ax.set_xticks(depths)
    ax.legend()

# fitting trees of depth 1 to 24
sm_tree_depths = range(1,25)
sm_cv_scores_mean, sm_cv_scores_std, sm_accuracy_scores = run_cross_validation_on_trees(X_train, y_train, sm_tree_depths)

# plotting accuracy
plot_cross_validation_on_trees(sm_tree_depths, sm_cv_scores_mean, sm_cv_scores_std, sm_accuracy_scores, 
                               'Accuracy per decision tree depth on training data')


In [ ]:
#printing out the optimum depth

idx_max = sm_cv_scores_mean.argmax()
sm_best_tree_depth = sm_tree_depths[idx_max]
sm_best_tree_cv_score = sm_cv_scores_mean[idx_max]
sm_best_tree_cv_score_std = sm_cv_scores_std[idx_max]
print('The depth-{} tree achieves the best mean cross-validation accuracy {} +/- {}% on training dataset'.format(
      sm_best_tree_depth, round(sm_best_tree_cv_score*100,5), round(sm_best_tree_cv_score_std*100, 5)))

The method selects tree depth 9-10 because it achieves the best average accuracy on training data using cross-validation folds with size 5. The lower bound of the confidence interval of the accuracy is high enough to make this value significant. When more nodes are added to the tree, it is clear that the cross-validation accuracy changes towards zero.

#### Now we print the confusion matrix

In [ ]:
# Extract Response and Predictors
y = pd.DataFrame(music['music_genre'])
X = pd.DataFrame(music[predictors])
# Split the Dataset into Train and Test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25) 
# Decision Tree using Train Data
dectree = DecisionTreeClassifier(max_depth = 10)  # create the decision tree object
dectree.fit(X_train, y_train)                    # train the decision tree model

# Predict Response corresponding to Predictors
y_train_pred = dectree.predict(X_train)
y_test_pred = dectree.predict(X_test)

# Check the Goodness of Fit (on Train Data)
print("Goodness of Fit of Model \tTrain Dataset")
print("Classification Accuracy \t:", dectree.score(X_train, y_train))
print()

# Check the Goodness of Fit (on Test Data)
print("Goodness of Fit of Model \tTest Dataset")
print("Classification Accuracy \t:", dectree.score(X_test, y_test))
print()

# Plot the Confusion Matrix for Train and Test
f, axes = plt.subplots(1, 2, figsize=(25, 10))
sns.heatmap(confusion_matrix(y_train, y_train_pred),
           annot = True, fmt=".0f", annot_kws={"size": 10}, ax = axes[0])
sns.heatmap(confusion_matrix(y_test, y_test_pred), 
           annot = True, fmt=".0f", annot_kws={"size": 10}, ax = axes[1])

In the above confusion matrix, 0-9 are the indexes for ["Classical","Jazz","Country","Rock","Hip-Hop","Rap","Anime","Alternative","Blues","Electronic"] respectively

The "Goodness of Fit" refers to the performance of the model, which has been evaluated using the accuracy metric. The accuracy metric measures the proportion of correctly classified instances out of all instances.

In this case, the model was trained on a training dataset and tested on a separate testing dataset. The classification accuracy for the training dataset was found to be 0.57 to 0.60 depending on the depth of tree, which means that the model correctly classified 57% to 60% of the instances in the training dataset. The classification accuracy for the testing dataset was found to be around 0.51, which means that the model correctly classified 51% of the instances in the testing dataset.

This shows that the decision tree may be an acceptable model for predicting song genres

# Second Machine Learning Model: Random Forest classifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, ConfusionMatrixDisplay
from sklearn.model_selection import RandomizedSearchCV, train_test_split
from scipy.stats import randint

Import the necessary functions for Random Forest model.

In [ ]:

rf = RandomForestClassifier(random_state=50,n_estimators=5,max_depth=4,min_samples_leaf=1)
rf.fit(X_train, y_train.values.ravel())

Do the necessary imports, spliting data into train and test set. Setting up Randomforestclassifier with preset "Random_state", "N-estimators=5" (Number of trees in the forest), "Max_depth=4" (Depth of the tree), "min_samples_leaf =1" (min leaf to determine the split). Fitting the train values into the model.

Random_State: Control the randomness involved in machine learning models
n_estimators: Number of trees to build in the forest.
max_depth: The depth of a tree.
min_samples_leaf : Minimum number of samples required to be at a leaf node.

In [ ]:
y_train_pred = rf.predict(X_train)
y_test_pred = rf.predict(X_test)

# Check the Goodness of Fit (on Train Data)
print("Goodness of Fit of Model \tTrain Dataset")
print("Classification Accuracy \t:", accuracy_score(y_train, y_train_pred))
print()

# Check the Goodness of Fit (on Test Data)
print("Goodness of Fit of Model \tTest Dataset")
print("Classification Accuracy \t:", accuracy_score(y_test, y_test_pred))
print()

# Plot the Confusion Matrix for Train and Test
f, axes = plt.subplots(1, 2, figsize=(25, 10))
sns.heatmap(confusion_matrix(y_train, y_train_pred),
           annot = True, fmt=".0f", annot_kws={"size": 10}, ax = axes[0])
sns.heatmap(confusion_matrix(y_test, y_test_pred), 
           annot = True, fmt=".0f", annot_kws={"size": 10}, ax = axes[1])

Using random forest classifier for the data training, with some pre-defined parameter returns both train and test classification accuracy of 0.47. which means that the model correctly classified 47% of the instances in both training and testing dataset. To further improve our modelling, we went on to implement GridSearch CV to improve our model.


## Implementing GridSearchCV


In [ ]:
from sklearn.model_selection import GridSearchCV

#### The code below takes around 12minutes to run, since GridSearchCV fit algorithm is O(n^2) at best, so it is extremely slow. Hyper parameters tuning.

In [ ]:
grid_param = {
    "n_estimators": [10, 15, 20, 25, 30, 35],
    "max_depth": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    "min_samples_leaf": [1, 2, 3, 4, 5]
}
grid_search = GridSearchCV(RandomForestClassifier(),param_grid=grid_param,scoring="accuracy",cv=5)
grid_search.fit(X_train,y_train.values.ravel())

In the grid_param {} each dictionary in the list will be used to explore. GridSearchCV for hyperparameter tuning in order to locate the best hyperparameters and improve model performance. In the function we predefined a list of values in grid_param for max_depth, n_estimators and minimum sample leaf to be looped through and fit the model on our training set.

In [ ]:
grid_search.best_params_

Retrieving the best parameters for model learning.

In [ ]:
rf3 = RandomForestClassifier(n_estimators = grid_search.best_params_['n_estimators'], max_depth = grid_search.best_params_['max_depth'], min_samples_leaf = grid_search.best_params_['min_samples_leaf'])
rf3.fit(X_train, y_train.values.ravel())

Fitting the best parameters into the classifier.

In [ ]:
y_train_pred3 = rf3.predict(X_train)
y_test_pred3 = rf3.predict(X_test)

# Check the Goodness of Fit (on Train Data)
print("Goodness of Fit of Model \tTrain Dataset")
print("Classification Accuracy \t:", accuracy_score(y_train, y_train_pred3))
print()

# Check the Goodness of Fit (on Test Data)
print("Goodness of Fit of Model \tTest Dataset")
print("Classification Accuracy \t:", accuracy_score(y_test, y_test_pred3))
print()

# Plot the Confusion Matrix for Train and Test
f, axes = plt.subplots(1, 2, figsize=(25, 10))
sns.heatmap(confusion_matrix(y_train, y_train_pred3),
           annot = True, fmt=".0f", annot_kws={"size": 10}, ax = axes[0])
sns.heatmap(confusion_matrix(y_test, y_test_pred3), 
           annot = True, fmt=".0f", annot_kws={"size": 10}, ax = axes[1])

Using the GridSearch CV does returns a better classification accuracy. Random forest classifier generated a better outcome than the multi-variate decision tree after applying both models to our dataset. Therefore, the best method would be to forecast music genres using random forest classifiers. 

Although the outcome might not be the best, 0.66 and 0.56 accuracy are regarded as relatively decent in predicting the music genres based on audio attributes given that it comprises numerous classes and does not have a strong association between each variable and its response "genre". Hence, based on the research and findings we have done, it is possible to predict the genre of a song using its auditory characteristics. However, there is more to be explored.